### PR TITLE
Improve error handling

### LINKED ISSUE #14: Why are we using `i16` for errors?
In `reactor::Error::TransportPollError` for example, and several other variants, we use `i16` for errors, instead of `io::Error`.

### LINKED ISSUE #15: Tidy up error variants
There are currently too many variants in `reactor::Error`, which makes the crate very hard to use, as it's unclear how each of these errors should be handled. Ideally most of the variants have clear actionables, eg. unregistering the socket. Here's what I'd start with.

---

These errors come from improper use of the API by the user, and shouldn't be reactor errors.
If we're trying to unregister an unknown peer, not returning an error makes unregistering idempotent,
which makes the API easier to use. If we're trying to write to an unknown peer, we could return a
`WriteFailure` error instead. Since they aren't fatal and can easily happen due to race conditions, I'd simply log an error
and remove the variants:

    reactor::Error::ListenerUnknown(id)
    reactor::Error::TransportUnknown(id)
    reactor::Error::WriteLogicError(id, _)

These are pretty critical and it's not clear how they can happen, but if our listener socket goes
down, we probably want to restart the node. I would probably merge the two variants into `Error::ListenerFailed`,
since there isn't really a difference in how they should be handled.

    reactor::Error::ListenerPollError(id, err)
    reactor::Error::ListenerDisconnect(id, _, err)

This error should be whatever the internal `Poll` implementation returns. Ideally it can be downcasted to the
concrete error type. Poll errors should be fatal, so we'd probably abort here.

    reactor::Error::Poll(err)

I'm guessing this one is a clean disconnect, in which case we can handle it and unregister the socket:

    reactor::Error::TransportDisconnect(id, _, err)

These two seem like they can be bundled under `Error::Transport`, since they'd likely both be handled by
disconnecting:

    reactor::Error::TransportPollError(id, err)
    reactor::Error::WriteFailure(id, err)

### PR DESCRIPTION
**Pull request by dr-orlovsky**
Closes #14, #15

### CODE REVIEW

File: src/reactor.rs [hunk 1]
```diff
@@ -701,54 +685,48 @@ impl<H: Handler, P: Poll> Runtime<H, P> {
...
+                             read-only or not ready for a write operation"
+                        );
                     }
-                    WriteError::Io(e) => {
+                    Err(WriteError::Io(e)) => {
                         #[cfg(feature = "log")]
-                        log::error!(target: "reactor", "Error writing to transport {id}: {e:?}");
-                        Error::WriteFailure(id, e)
+                        log::error!(target: "reactor", "Fatal error writing to transport {id}, disconnecting. Error details: {e:?}");
+                        self.unregister_transport(id);
```
Comment (cloudhead) [inline src/reactor.rs:726]: Do I receive an error when this happens though? I'd expect a `TransportDisconnect` error probably, and *not* a call to `handover_transport` without error.

File: src/resource.rs [hunk 2]
```diff
@@ -88,11 +89,27 @@ pub enum WriteError {
...
+    /// # Safety
+    ///
+    /// Panics on invalid [`WriteAtomic::write_or_buf`] implementation, i.e. if it doesn't handle
+    /// EGAGAIN, EINTER, EWOULDBLOCK I/O errors by buffering the data and returns them instead.
     fn write_atomic(&mut self, buf: &[u8]) -> Result<(), WriteError> {
         if !self.is_ready_to_write() {
             Err(WriteError::NotReady)
         } else {
-            self.write_or_buf(buf).map_err(WriteError::from)
+            // TODO: on EGAGAIN, EINTER, EWOULDBLOCK just keep the data buffered
```
Comment (cloudhead) [inline src/resource.rs:101]: Is it a problem that we don't do this atm?

By the way `EINTR` can be retried immediately (in a loop). That's how `io::copy` works for eg.

File: src/resource.rs [hunk 3]
```diff
@@ -113,5 +130,11 @@ pub trait WriteAtomic: io::Write {
     /// with a system-level error.
     ///
     /// This method shouldn't be called directly; [`Self::write_atomic`] must be used instead.
+    ///
+    /// # Safety
+    ///
+    /// The method must handle EGAGAIN, EINTER, EWOULDBLOCK I/O errors and buffer the data in such
+    /// cases. Ig these errors are returned from this methods [`WriteAtomic::write_atomic`] will
+    /// panic.
```
Comment (cloudhead) [inline src/resource.rs:138]: Interesting. You don't want to handle those as regular errors? I guess it only panics in `debug` mode according to the above `debug_assert`? Should that be specified here?

**Comment by cloudhead** [comment]
Glad you're working on this. Looks good overall, just some questions :+1:

File: src/resource.rs (hunk 2 above)
Comment (dr-orlovsky) [reply src/resource.rs:101]: This is not a write operation, but a buffered write, meaning that it is the job of the resource to buffer on such errors and not to propagate them (it is the current behavior we have in `netservices`

File: src/resource.rs (hunk 3 above)
Comment (dr-orlovsky) [reply src/resource.rs:138]: The reply to the comment above applies here as well

File: src/reactor.rs (hunk 1 above)
Comment (dr-orlovsky) [reply src/reactor.rs]: Updated in the pushed commit. I was under an impression there is no need to handover the transport (which is a part of the error message) to the service. And w/o handover there were no need in removing the previous version of the error.

**Review by cloudhead — APPROVED** [review]
LGTM

**Pull request merged**